.gnu-emacs/config.org
#+TITLE: Gnu Emacs Config
#+AUTHOR: HEWA S.

* Table of Content :toc:
- [[#about][About]]
  - [[#important][Important]]
- [[#packageel][Package.el]]
  - [[#repo][Repo]]
  - [[#use-package][use-package]]
  - [[#ensure-system-package][ensure system package]]
  - [[#delight][delight]]
- [[#garbage-collector-magic-hack][Garbage Collector Magic Hack]]
- [[#startup-and-ui][Startup and UI]]
- [[#appearance][Appearance]]
  - [[#themes][Themes]]
  - [[#doom-moodline][Doom-Moodline]]
  - [[#whichkey][Whichkey]]
  - [[#emojies][Emojies]]
  - [[#icons][Icons]]
  - [[#rainbow-mode][Rainbow-mode]]
- [[#generalel][General.el]]
- [[#evil][Evil]]
- [[#magit][Magit]]
- [[#org][ORG]]
  - [[#org-mode][ORG Mode]]
  - [[#org-bullets][Org-Bullets]]
  - [[#get-the-old-easy-template-system-back-snippet][Get the old easy template system back "snippet"]]
  - [[#automatically-create-table-of-contents][Automatically Create Table of Contents]]
- [[#ivy][Ivy]]
  - [[#swiper][Swiper]]
  - [[#counsel][Counsel]]
  - [[#ivy-1][Ivy]]
- [[#smex][Smex]]
- [[#smartparens][Smartparens]]
- [[#company][Company]]
- [[#syntax-spell-checker][Syntax, Spell-Checker]]
  - [[#flyc][Flyc]]
  - [[#spell-checker][Spell-Checker]]
- [[#simpleclip][Simpleclip]]
- [[#development][Development]]
  - [[#languages][Languages]]
  - [[#yasnippet][YASnippet]]
- [[#dashboard][Dashboard!]]
- [[#elfeed][Elfeed]]
- [[#dummy-packages][Dummy Packages!]]
- [[#projectile][Projectile]]

* About
This is my Emacs config inspired by Doom-Emacs.
** Important
This is a literate configuration for Emacs written in Org format.
The following code block should be your init.el.  
This tells the Emacs to use the /source code blocks/ from:
+ config.org global
+ keybinding.org global
+ linux.org specific config for Gnu/Linux
+ darwin.org specific config for OSX
+ helpers.org functions, specific code for testing
#+begin_example
;; global configuration
(org-babel-load-file
 (expand-file-name "config.org" user-emacs-directory))

(org-babel-load-file
 (expand-file-name "keybinding.org" user-emacs-directory))

(org-babel-load-file
 (expand-file-name "helpers.org" user-emacs-directory))

;; check operating system
;; use ‘C-h v system-type’ (command ‘describe-variable’) to see the current value of ‘system-type’
(cond ((eq system-type 'gnu/linux)
       ;; Linux-specific code goes here.
       (org-babel-load-file 
         (expand-file-name "linux.org" user-emacs-directory))
       )

      ((eq system-type 'darwin)
       ;; OSX-specific code goes here.
       (org-babel-load-file
         (expand-file-name "darwin.org" user-emacs-directory))
       ))

;; EOF
;; EOF
;; EOF
#+end_example

* Package.el
** Repo
Get packages from the following sources.
#+begin_src emacs-lisp
(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
(add-to-list 'package-archives '("org" . "https://orgmode.org/elpa/") t)
#+end_src

** use-package
 The use-package macro allows you to isolate package configuration
 in your .emacs file in a way that is both performance-oriented and, well, tidy.
#+begin_src emacs-lisp
(unless package-archive-contents    ;; Refresh the packages descriptions
  (package-refresh-contents))
(setq package-load-list '(all))     ;; List of packages to load
(unless (package-installed-p 'use-package)  ;; Make sure the use-package package is
  (package-install 'use-package))           ;; installed, install it if not
(package-initialize)                ;; Initialize & Install Package

(require 'use-package-ensure)
(setq use-package-always-ensure t) ;; Global
;; NOTE: :ensure will install a package if it is not already installed,
;; but it does not keep it up-to-date. If you want to keep your packages updated automatically,
;; one option is to use auto-package-update, 
#+end_src

** ensure system package
 The :ensure-system-package keyword allows you
 to ensure system binaries exist alongside your package declarations.
#+begin_src emacs-lisp
  (use-package use-package-ensure-system-package)
#+end_src

** delight
Enables you to customise the mode names displayed in the mode line.
more info: https://elpa.gnu.org/packages/delight.html
#+begin_src emacs-lisp
(use-package delight)
#+end_src

* Garbage Collector Magic Hack
#+begin_src emacs-lisp
(use-package gcmh
  :custom
  (gcmh-idle-delay 0.5)  ; default is 15s
  (gcmh-high-cons-threshold (* 16 1024 1024))  ; 16mb
  :init
  (gcmh-mode))

#+end_src
* Startup and UI
#+begin_src emacs-lisp
;; Mouse interface 
(tool-bar-mode -1)                                ; Hide toolbar
(scroll-bar-mode -1)                              ; Hide scrollbar
(delete-selection-mode t)                         ; Enable selection via mouse
;; Line numbers and Truncated lines
(global-display-line-numbers-mode 1)              ; Display line numbers
(global-visual-line-mode t)                       ; Allways display long lines
(column-number-mode 1)                            ; Show the column number
(global-hl-line-mode)                             ; Hightlight current line
(prefer-coding-system 'utf-8-unix)                ; Default to utf-8 unix encoding
;; Yes or no
(fset 'yes-or-no-p 'y-or-n-p)                     ; Replace yes/no prompts with y/n
(set-default-coding-systems 'utf-8)               ; Default to utf-8 encoding
(show-paren-mode 1)                               ; Show the parent, start ( end )
(setq sentence-end-double-space nil)              ; Sentence should end with only a point.
(setq require-final-newline t)                    ; Newline at end of file
(setq vc-follow-symlinks t )                      ; Don't ask for confirmation when opening symlinked file
;; Better defaults
(setq-default
 cursor-in-non-selected-windows t                 ; Hide the cursor in inactive windows
 fill-column 80                                   ; Set width for automatic line breaks
 help-window-select t                             ; Focus new help windows when opened
 inhibit-startup-screen t                         ; Disable start-up screen
 initial-scratch-message ";;Welcome Back!"          ; Default message in the initial *scratch* buffer
 kill-ring-max 128                                ; Maximum length of kill ring, default is 60
 load-prefer-newer t                              ; Prefers the newest version of a file
 mark-ring-max 128                                ; Maximum length of mark ring
 read-process-output-max (* 1024 1024)            ; Increase the amount of data reads from the process, default is 4k
 scroll-conservatively most-positive-fixnum       ; Always scroll by one line
 select-enable-clipboard t                        ; Merge system's and Emacs' clipboard
 indent-tabs-mode nil                             ; Prefers spaces over tabs
 tab-width 4                                      ; Set width for tabs
 user-full-name "Hewa S."                         ; Set the full name of the current user
 user-mail-address "hewa9g[at]gmail.com"          ; Set the email address of the current user
 vc-follow-symlinks t                             ; Always follow the symlinks
; debug-on-error t                                 ; Debug on error
 view-read-only t)                                ; Always open read-only buffers in view-mode
;; Backup files
(setq indent-line-function 'insert-tab)           ; Tab inserts 4 space 
(setq vc-make-backup-files t )		            ; Backup file even when in version controlled dir
(setq backup-directory-alist `(("." . "~/.cache/emacs/backups")) ) ; Backup files to ~/.cache/emacs/backups
(setq auto-save-file-name-transforms '((".*" "~/.cache/emacs/auto-save-list/" t)) ) ; Transform backup file names
(setq delete-old-versions -1)                     ; Don't delete backup files silently
(setq ad-redefinition-action 'accept)             ; Hide warnings from defadvice
;;(setq gc-cons-threshold 100000000)                ;(100mb), The default setting is too low for lsp-mode
#+end_src
* Appearance
** Themes
#+begin_src emacs-lisp
(use-package doom-themes)
;; Global settings (defaults)
(setq doom-themes-enable-bold t    ; if nil, bold is universally disabled
      doom-themes-enable-italic t) ; if nil, italics is universally disabled

;; Load the theme (doom-one, doom-molokai, etc); keep in mind that each
;; theme may have their own settings.
(load-theme 'doom-one t)

;; Enable flashing mode-line on errors
;; (doom-themes-visual-bell-config)
#+end_src

** Doom-Moodline
#+begin_src emacs-lisp
(use-package doom-modeline
  :init
  (doom-modeline-mode 1)
  :custom
  ;; If non-nil, a word count will be added to the selection-info modeline segment.
  (doom-modeline-enable-word-count t)
  ;; Whether display the indentation information.
  (doom-modeline-indent-info t)
  )
#+end_src
** Whichkey
Which-Key is a minor mode for Emacs that displays the key bindings following your currently entered incomplete command (a prefix) in a popup
#+begin_src emacs-lisp
(use-package which-key

  :init
  (setq which-key-side-window-location 'bottom
        which-key-sort-order #'which-key-key-order-alpha
        which-key-sort-uppercase-first nil
        which-key-add-column-padding 1
        which-key-max-display-columns nil
        which-key-min-display-lines 6
        which-key-side-window-slot -10
        which-key-side-window-max-height 0.75
        which-key-idle-delay 0.8
        which-key-max-description-length 35
        ;;which-key-popup-type 'minibuffer              ;default: side-window, other:fram or custom
        ;;which-key-allow-imprecise-window-fit t
        which-key-separator " → " )
;; Hide mode-line
  (add-hook 'which-key-init-buffer-hook (lambda () (setq-local mode-line-format nil)))
  )
(which-key-mode)
#+end_src

** Emojies
Emojify is an Emacs extension to display emojis. It can display github style emojis like :smile: or plain ascii ones like :).
#+begin_src emacs-lisp
(use-package emojify
  :hook (after-init . global-emojify-mode))
#+end_src

** Icons
A library for inserting Developer icons.
#+begin_src emacs-lisp
(use-package all-the-icons)
(use-package all-the-icons-dired
  :hook (dired-mode . all-the-icons-dired-mode))
#+end_src

** Rainbow-mode
Colorize color names in buffers. This is "#FFFF00" yellow :)
#+begin_src emacs-lisp
;;The :hook keyword allows adding functions onto package hooks
(use-package rainbow-mode
;;  :delight ;; Don't show anything for rainbow-mode.
:hook
(org-mode python-mode) ;; apply for multiple mode
;;((org-mode python-mode) . rainbow-mode );; another syntax
  )
#+end_src
* General.el 
NOTE: load general before evil, otherwise the :general keyword in the use-package blocks won't work.

/general.el/ provides a more convenient method for binding keys in emacs (for both evil and non-evil users). 
#+begin_src emacs-lisp
(use-package general
  :after whick-key
  :config
  (setq general-override-states '(normal visual insert emacs hybrid motion operator replace))
  ;;    (general-override-mode 1) this is default now
  )
(general-def
  :states '(normal visual insert emacs)
  :keymaps 'override ;;;; to prevent your leader keybindings from ever being overridden (e.g. an evil
  "M-x" '(counsel-M-x  :which-key "Ivy version of execute-extended-command")
  "C-s" '(swiper  :which-key "an alternative to isearch")
  ;; zoom in/out like we do everywhere else.
  "C-=" '(text-scale-increase :which-key "zoom in")
  "C--" '(text-scale-decrease :which-key "zoom out")
  )
;; simple function to change current state in all buffers. OBSOLETE
;;(with-current-buffer (buffer-name)
;;  (evil-normal-state))

;; custom keybinding are defined under keybinding.org with general-create-definer
#+end_src
* Evil
Evil is an extensible vi layer for Emacs. It emulates the main features of Vim, and provides facilities for writing custom extensions.
#+begin_src emacs-lisp
(use-package evil
  :after general ; optional, since general loaded befor evil
  :init
  ;; integrate with evil-collection
  (setq evil-want-integration t) ;; This is optional since it's already set to t by default.
  (setq evil-want-keybinding nil)
 
  (add-hook 'with-editor-mode-hook 'evil-insert-state) ;;set magit-commit in insert mode
  :config
  (setq evil-vsplit-window-right t)
  (setq evil-split-window-below t)
  (evil-mode 1))

(use-package evil-collection
  :after evil
  :config
  ;;  (setq evil-collection-mode-list '(dashboard dired ibuffer magit))
  (evil-collection-init))
#+end_src
* Magit
Magit is a complete text-based user interface to Git.
#+begin_src emacs-lisp
(use-package magit)
#+end_src

* ORG
** ORG Mode
#+begin_src emacs-lisp
(use-package org
  :delight "ئۆڕگ"
  :init
  (setq org-startup-folded 'content) ;fold headlines and parents at startup
  :hook 
  ;; Wrap long lines while in org mode
  (org-mode . visual-line-mode)
  ;; Indent content under headers, which makes for easier reading
  (org-mode . org-indent-mode) 
  :config
  ;; Run/highlight code using babel in org-mode
  (org-babel-do-load-languages
   'org-babel-load-languages
   '(
     (emacs-lisp . t)
     (org . t)
     (shell . t)
     (C . t)
     (python . t)
     (gnuplot . t)
     (octave . t)
     (R . t)
     (dot . t)
     (awk . t)
 ;;    (ipython . t)
     ;; Include other languages here...
     ))

  :custom
  (org-indent-indentation-per-level 2)
  (org-src-tab-acts-natively t)
  (org-edit-src-content-indentation 0)
  (org-src-fontify-natively t) ;; Syntax highlight in #+BEGIN_SRC blocks
  (org-confirm-babel-evaluate nil) ;; Don't prompt before running code in org
  (org-src-preserve-indentation t)
;;  (ob-async-no-async-languages-alist '("ipython")) ;; Fix an incompatibility between the ob-async and ob-ipython packages
  ;;  (electric-indent-mode -1)


  )
#+end_src

** Org-Bullets
Show bullets in org-mode as UTF-8 characters
#+begin_src emacs-lisp
(use-package org-bullets
  :custom
  ;; org-bullets-bullet-list
  ;; to reduse startup time, don't use emojies HSK
  (org-bullets-bullet-list '("✿" "◉" "◐" "►"))
  (org-ellipsis "▼")
  :hook
  (org-mode . org-bullets-mode))
#+end_src

** Get the old easy template system back "snippet"
#+begin_src emacs-lisp
(use-package org-tempo
  :ensure nil) ;; tell use-package not to try to install org-tempo since it's already there.
#+end_src
** Automatically Create Table of Contents
#+begin_src emacs-lisp
(use-package toc-org
  :commands toc-org-enable
;;come back later to use :hook instead of :init
  :init (add-hook 'org-mode-hook 'toc-org-enable))
#+end_src
* Ivy
Ivy is a generic completion mechanism for Emacs. While it operates similarly to other completion schemes such as icomplete-mode, Ivy aims to be more efficient, smaller, simpler, and smoother to use yet highly customizable.
Ivy re-uses the following packages if they are installed: avy, amx or smex, flx, and wgrep.
** Swiper
Swiper is an alternative to isearch that uses Ivy to show an overview of all matches.
#+begin_src emacs-lisp
(use-package swiper
  :after ivy
  :general
  ;; uses `general-def' not `general-define-key', so this is fine
  (swiper-map "M-%" 'swiper-query-replace)
  )
#+end_src

** Counsel
a collection of Ivy-enhanced versions of common Emacs commands
Enabling counsel-mode remaps built-in Emacs functions that have counsel replacements:
| Emacs command            | Counsel equivalent         |
|--------------------------+----------------------------|
| execute-extended-command | counsel-M-x                |
| describe-bindings        | counsel-descbinds          |
| describe-function        | counsel-describe-function  |
| describe-variable        | counsel-describe-variable  |
| apropos-command          | counsel-apropos            |
| describe-face            | counsel-describe-face      |
| list-faces-display       | counsel-faces              |
| find-file                | counsel-find-file          |
| find-library             | counsel-find-library       |
| imenu                    | counsel-imenu              |
| load-library             | counsel-load-library       |
| load-theme               | counsel-load-theme         |
| yank-pop                 | counsel-yank-pop           |
| info-lookup-symbol       | counsel-info-lookup-symbol |
| pop-to-mark-command      | counsel-mark-ring          |
| bookmark-jump            | counsel-bookmark           |

#+begin_src emacs-lisp
(use-package counsel
  :after ivy
  :config (counsel-mode)
  )
#+end_src

** Ivy
#+begin_src emacs-lisp
(use-package ivy
  :general
  (ivy-minibuffer-map
   "TAB" 'ivy-partial               ;; Don't open files, i use RET
   "C-j" 'ivy-next-line
   "C-k" 'ivy-previous-line
   "C-l" 'ivy-alt-done)
  (ivy-switch-buffer-map
   "C-j" 'ivy-next-line
   "C-k" 'ivy-previous-line
   "C-;" 'ivy-switch-buffer-kill)
  :defer 0.1
  :delight
  :after ivy-rich
  :custom
  (ivy-case-fold-search-default t) ;;allways use case-insensitive
  (ivy-count-format "(%d/%d) ")
  (enable-recursive-minibuffers t)
  (ivy-use-virtual-buffers t)
  (ivy-initial-inputs-alist nil) ;; hide "^" from M-x
  :config (ivy-mode)
  )
;; This package comes with rich transformers for commands from ivy and counsel
(use-package ivy-rich
  :config
  (ivy-rich-mode)
  )
#+end_src

* Smex
Provide a convenient interface to recently and most frequently used commands.
#+begin_src emacs-lisp
(use-package smex
  :config
  (smex-initialize)
  )
#+end_src

* Smartparens
Smartparens is a minor mode for dealing with pairs in Emacs.
It can automatically insert pairs.
#+begin_src emacs-lisp
(use-package smartparens
  :defer 1
  :delight
  :custom (sp-escape-quotes-after-insert nil)
  :config (smartparens-global-mode 1))

#+end_src
* Company
#+begin_src emacs-lisp
(use-package company
  :general
  (company-active-map
   "C-j" 'company-select-next
   "C-k" 'company-select-previous
   "C-l" 'company-complete-selection
   )
  :defer 0.5
  :delight
  :custom
  (company-begin-commands '(self-insert-command))
  (company-idle-delay 0.2)
  (company-minimum-prefix-length 2)
  (company-show-numbers t)
  (company-tooltip-align-annotations 't)
  ;;(global-company-mode t)
  )

;; company-box take care of documentation popups as well, don't use company-quickhelp!
(use-package company-box
  :after company
  :delight
  :hook (company-mode . company-box-mode))
#+end_src
* Syntax, Spell-Checker
** Flyc
Flycheck is a modern on-the-fly syntax checking extension for GNU Emacs, intended as replacement for the older Flymake extension which is part of GNU Emacs.
#+begin_src emacs-lisp
;; M-x flycheck-verify-setup
(use-package flycheck
  :init
  (global-flycheck-mode)
  )
#+end_src

** Spell-Checker
check this [[https://200ok.ch/posts/2020-08-22_setting_up_spell_checking_with_multiple_dictionaries.htmli][article]] , use Hunspell
apt install hunspell hunspell-de-de hunspell-en-gb hunspell-en-us hunspell-de-ch-frami
#+begin_src emacs-lisp
(with-eval-after-load "ispell"
  ;; Configure `LANG`, otherwise ispell.el cannot find a 'default
  ;; dictionary' even though multiple dictionaries will be configured
  ;; in next line.
  (setenv "LANG" "en_US.UTF-8")
  (setq ispell-program-name "hunspell")
  ;; Configure German, Swiss German, and two variants of English.
  (setq ispell-dictionary "de_DE,de_CH,en_GB,en_US")
  ;; ispell-set-spellchecker-params has to be called
  ;; before ispell-hunspell-add-multi-dic will work
  (ispell-set-spellchecker-params)
  (ispell-hunspell-add-multi-dic "de_DE,de_CH,en_GB,en_US")
  ;; For saving words to the personal dictionary, don't infer it from
  ;; the locale, otherwise it would save to ~/.hunspell_de_DE.
  (setq ispell-personal-dictionary "~/.hunspell_personal"))

;; The personal dictionary file has to exist, otherwise hunspell will
;; silently not use it.
(unless (file-exists-p ispell-personal-dictionary)
  (write-region "" nil ispell-personal-dictionary nil 0))
#+end_src

* Simpleclip
Simplified access to the system clipboard in Emacs.
#+begin_src emacs-lisp
(use-package simpleclip
  :general
  ("C-S-c" 'simpleclip-copy)
;;  :init
;;  (simpleclip-mode 1)
  )
#+end_src
* Development
** Languages
*** Python
#+begin_src emacs-lisp
(use-package python-mode
  :delight "پایتۆن"
  :hook
  ;;  (python-mode . pyvenv-mode)
  (python-mode . flycheck-mode)
  (python-mode . company-mode)
  (python-mode . elpy-enable)
  ;;  (python-mode . blacken-mode)
  )

;; sudo apt-get install flake8
;;TODO: add fold/unfold Keybinding
(use-package elpy
  :after python-mode
  :custom
  (elpy-rpc-virtualenv-path "~/PythonVenv")
  (pyvenv-activate "~/PythonVenv")
  ;;  :init
  ;;  (elpy-enable)
  :config
  ;; elpy uses built-in flymake by default!
  (setq elpy-modules (delq 'elpy-module-flymake elpy-modules)) ;; remove flymake and use flycheck!
  ;;  (setq python-shell-interpreter "ipython"
  ;;        python-shell-interpreter-args "-i --simple-prompt")
  )
#+end_src

*** PATH
#+begin_src emacs-lisp
(use-package exec-path-from-shell
  :if (memq window-system '(mac ns x))
  :config
  (exec-path-from-shell-initialize))
#+end_src

** YASnippet
Template system for Emacs.
Note: YASnippet no longer bundles snippets directly!
use yasnippet-snippets, which is a snippet collection package.
M-x yas-describe-tables : Display Snippets and Keys
#+begin_src emacs-lisp
(use-package yasnippet
:config
(yas-global-mode))

(use-package yasnippet-snippets
  :after yasnippet
  :config (yasnippet-snippets-initialize))
#+end_src

* Dashboard! 
#+begin_src emacs-lisp
(use-package dashboard
  :custom-face

.config/qtile/config.py
# -*- coding: utf-8 -*-
# Compatable with Qtile version : "0.18.0 Stable" or later.
# Modified by : Hewa S.

# Copyright (c) 2010 Aldo Cortesi
# Copyright (c) 2010, 2014 dequis
# Copyright (c) 2012 Randall Ma
# Copyright (c) 2012-2014 Tycho Andersen
# Copyright (c) 2012 Craig Barnes
# Copyright (c) 2013 horsik
# Copyright (c) 2013 Tao Sauvage

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

from typing import List  # noqa: F401

# mouse_callbacks
from libqtile import qtile

#----- Built-in Extensions: WindowList
from libqtile.extension import WindowList

import os, subprocess, socket
from libqtile import bar, layout, hook
from libqtile.config import Click, Drag, Group, Key, Match, Screen, KeyChord
from libqtile.lazy import lazy
from libqtile.utils import guess_terminal
from libqtile.widget import (
    CurrentLayoutIcon,
    Sep,
    GroupBox,
    TextBox,
    Prompt,
    WindowName,
    Systray,
    Net,
    Volume,
    Clock,
    KeyboardLayout,
    WindowCount,
#    CapsNumLockIndicator,
#    Chord,
    Memory,
    CPU,
    QuickExit,
    )

# DEBUG: uncomment this section
#try:
#    import aiomanhole
#except ImportError:
#    aiomanhole = None

# DEBUG = os.environ.get("DEBUG")

mod = "mod4"
terminal = guess_terminal()

groups = [
    Group("MAX", layout='max', matches=[Match(wm_class=["Google-chrome"])]),
    Group("DEV", layout='matrix', matches=[Match(wm_class=["pcmanfm"])]),
    Group("ORG", layout='max', matches=[Match(wm_class=["emacs"])]),
    Group("VLC", layout='max', matches=[Match(wm_class=["vlc"])]),
]

#----- Play around Groups
# MOD4 + index Number : Switch to Group[index]
# MOD4 + shift + index Number : Send active window to another Group
from libqtile.dgroups import simple_key_binder
dgroups_key_binder = simple_key_binder("mod4")

#----- Key Bindings
keys = [
    # Switch between windows
    Key([mod], "h", lazy.layout.left(), desc="move focus to left"),
    Key([mod], "l", lazy.layout.right(), desc="move focus to right"),
    Key([mod], "j", lazy.layout.down(), desc="move focus down"),
    Key([mod], "k", lazy.layout.up(), desc="move focus up"),
    Key([mod], "space", lazy.layout.next(),
        desc="Move window focus to other window"),

    # Move windows between left/right columns or move up/down in current stack.
    Key([mod, "shift"], "h", lazy.layout.swap_left(),
        desc="move active window >> left",),
    Key([mod, "shift"], "l", lazy.layout.swap_right(),
        desc="move active window >> right",),
    Key([mod, "shift"], "j", lazy.layout.shuffle_down(),
        desc="move active window >> down",),
    Key([mod, "shift"], "k", lazy.layout.shuffle_up(),
        desc="move active window >> up",),
    
    # Change window size/ratio
    Key([mod], "i", lazy.layout.grow(),
        desc="expand window size/ratio",),
    Key([mod], "m", lazy.layout.shrink(),
        desc="shrink/decrease window size/ratio",),
    Key([mod], "n", lazy.layout.normalize(),
        desc="normalize stack size/ratio",),
    Key([mod], "o", lazy.layout.maximize(),
        desc="maximize active window size/ratio : stack",),
    Key([mod], "c", lazy.layout.reset(),
        desc="reset all windows",),
    Key([mod, "shift"], "space", lazy.layout.flip(),
        desc="flip stack layout",),

    # Float active window or reset to the correct order
    Key([mod, "shift"], "f", lazy.window.toggle_floating(),
        desc="Toggle Floating"),

    # Applications
    Key([mod], "Return", lazy.spawn(terminal), desc="Launch terminal"),
    Key([mod], "p", lazy.spawn("dmenu_run -i"), desc="dmenu"),
    Key([mod, "shift"], "w", lazy.spawn("chromium"),
        desc="Launch Chromium Browser"),

    Key([mod, "shift"], "Return", lazy.widget["keyboardlayout"].next_keyboard(), desc="Next keyboard layout."),
    Key([mod], "e", lazy.spawn("emacsclient -c -a emacs"), desc="Emacs"),
    Key([mod], "d", lazy.spawn("emacs --with-profile doom"), desc="Doom-Emacs"),
    Key([mod], "v", lazy.spawn("emacs --with-profile vanilla"), desc="Vanilla-Emacs"),

    # Toggle between different layouts as defined below
    Key([mod], "Tab", lazy.next_layout(), desc="Toggle between layouts"),
    Key([mod], "w", lazy.window.kill(), desc="Kill focused window"),

    Key([mod, "control"], "r", lazy.restart(), desc="Restart Qtile"),
    Key([mod, "control"], "q", lazy.shutdown(), desc="Shutdown Qtile"),
    Key([mod], "r", lazy.spawncmd(),
        desc="Spawn a command using a prompt widget"),

    # List open windows in dmenu 
    Key([mod], "u", lazy.run_extension(WindowList(
        item_format="{id}: {window} >>> {group}",)),
        desc="Give vertical list of all open windows in dmenu. Switch to selected",
        ),
]

#----- GLOBAL COLORZ, Yep ColorZ!
colorz = ("#282c34", #[0] Dark Panel BG
          "#2ABB9B", #[1] Green : Border Line >> Active Tab >> Group
          "#FFa400", #[2] Yellow: Inactive Group Names
          "#F22613", #[3] Red   : Active Group Names
          "#19B5FE", #[4] Blue  : Selected Group
          "#ff0000", #[5] light Red : Focused Window Border
          "#5aff00", #[6] Green : Normal Window Border
          )

layout_monad = {
    "border_focus": colorz[5],
    "border_normal": colorz[6],
    "border_width": 2,
    "margin": 2,
    "single_border_width": 2,
    "single_margin": 2,
    }

layout_flo = {
    "border_focus": colorz[5],
    "border_normal": colorz[6],
    "border_width": 2,
    "fullscreen_border_width": 2,
    }

layout_matrix = {
    "border_focus": colorz[5],
    "border_normal": colorz[6],
    "border_width": 2,
    "margin": 2,
    }

layouts = [
    layout.MonadTall(**layout_monad,ratio=0.6),
    layout.MonadWide(**layout_monad),
    layout.Matrix(**layout_matrix),
    # layout.Zoomy(**layout_zoomy),
    layout.Max(),
    # layout.Columns(border_focus_stack='#d75f5f'),
    # Try more layouts by unleashing below layouts.
    # layout.Stack(num_stacks=2),
    # layout.Bsp(),
    # layout.RatioTile(),
    # layout.Tile(),
    # layout.TreeTab(),
    # layout.VerticalTile(),
]

#----- Widget Settings
widget_defaults = dict(
    font='Arimo Regular Nerd Font Complete',
    fontsize=13,
    padding=3,
    background=colorz[0], #Panel Background
    )

#----- Random Global Variables
#----- Custom Icon Path for CurrentlayoutIcon Widget
wd_icon = [os.path.expanduser("~/.config/qtile/icons")]

# Default File Manager
def open_fm():
    qtile.cmd_spawn('pcmanfm')

w_sep = {
    "background": colorz[0],
    "foreground": colorz[2],
    "linewidth":8,
    "size_percent":60,
    }

w_gbox = {
    "active": colorz[3],
    "block_highlight_text_color": colorz[4],
    "borderwidth": 2,
    "inactive": colorz[2],
    "this_current_screen_border": colorz[1],
    "this_screen_border": colorz[2],
    }

w_prompt = {
    "cursor_color": colorz[3],
    "padding": 5,
    "prompt": "Start:   ",
    }

w_net = {
    "format": "{down} ↓↑ {up}",
    "interface": "wlx7cdd90399231",
    }

w_vol = {
    "fmt": "🎧 {}",
    }

w_clock = {
    "format": "%d-%m-%Y %a %I:%M %p",
}

w_key_l = {
    "configured_keyboards": ['us','de','iq ku_ara', 'ar'],
    "display_map": {"us":"EN",
                    "iq ku_ara":"KU",
                    "de":"DE",
                    "ar":"AR"},
}

screens = [
    Screen(
        top=bar.Bar(
            [
                CurrentLayoutIcon(custom_icon_paths=wd_icon),
                Sep(**w_sep),
                GroupBox(**w_gbox,),
                TextBox(text="🐸", fontsize="16", padding=5,),
                WindowCount(),
                Prompt(**w_prompt),
                WindowName(foreground=colorz[1]),
                TextBox(text="📂", fontsize="16", padding=5, mouse_callbacks={'Button1': open_fm},),
                Systray(icon_size=20),
                TextBox(text= "[", foreground= colorz[6], fontsize= 18,),
                Memory(),
                TextBox(text= "]", foreground= colorz[6], fontsize= 18,),
                TextBox(text= "[", foreground= colorz[4], fontsize= 18,),
                CPU(),
                TextBox(text= "]", foreground= colorz[4], fontsize= 18,),
                TextBox(text="[", foreground=colorz[2], fontsize=18,),
                Net(**w_net,),
                TextBox(text="]", foreground=colorz[2], fontsize=18,),
                TextBox(text= "[", foreground= colorz[4], fontsize= 18,),
                Volume(**w_vol),
                TextBox(text= "]", foreground= colorz[4], fontsize= 18,),
                TextBox(text= "[", foreground= colorz[3], fontsize= 18,),
                Clock(**w_clock),
                TextBox(text= "]", foreground= colorz[3], fontsize= 18,),
                TextBox(text= "[", foreground= colorz[1], fontsize= 18,),
                KeyboardLayout(**w_key_l),
                TextBox(text= "]", foreground= colorz[1], fontsize= 18,),

            ],
            24,
            #opacity=0.90,
        ),
    ),
]

@hook.subscribe.startup_once
def autostart():
    home = os.path.expanduser('~/.config/qtile/autostart.sh')
    subprocess.call([home])

# DEBUG: uncomment this section
#if aiomanhole:
#    @hook.subscribe.startup_complete
#    def set_manhole():
#        aiomanhole.start_manhole(port=7113, namespace={"qtile": qtile})

#hostname = socket.gethostname()

mouse = [
    Drag([mod], "Button1", lazy.window.set_position_floating(),
         start=lazy.window.get_position()),
    Drag([mod], "Button3", lazy.window.set_size_floating(),
         start=lazy.window.get_size()),
    Click([mod], "Button2", lazy.window.bring_to_front())
]

#dgroups_key_binder = None #HSK
dgroups_app_rules = []  # type: List
follow_mouse_focus = True
bring_front_click = False
cursor_warp = False
floating_layout = layout.Floating(**layout_flo, float_rules=[
    # Run the utility of `xprop` to see the wm class and name of an X client.
    *layout.Floating.default_float_rules,
    Match(wm_class='confirmreset'),  # gitk
    Match(wm_class='makebranch'),  # gitk
    Match(wm_class='maketag'),  # gitk
    Match(wm_class='ssh-askpass'),  # ssh-askpass
    Match(title='branchdialog'),  # gitk
    Match(title='pinentry'),  # GPG key password entry
    Match(wm_class='nm-connection-editor'),  # Network manager 
])
auto_fullscreen = True
focus_on_window_activation = "smart"
reconfigure_screens = True

# If things like steam games want to auto-minimize themselves when losing
# focus, should we respect this or not?
auto_minimize = True

# XXX: Gasp! We're lying here. In fact, nobody really uses or cares about this
# string besides java UI toolkits; you can see several discussions on the
# mailing lists, GitHub issues, and other WM documentation that suggest setting
# this string if your java app doesn't work correctly. We may as well just lie
# and say that we're a working one by default.
#
# We choose LG3D to maximize irony: it is a 3D non-reparenting WM written in
# java that happens to be on java's whitelist.
wmname = "LG3D"
